File: 프로그래머스/lv3/43162. 네트워크/네트워크.java
import java.util.Stack;
 
class Solution {
    boolean[] visited;

    public int solution(int n, int[][] computers) {
        int answer = 0;
        visited = new boolean[n];
        
        Stack<Integer> stack = new Stack<>();
 
        for (int i = 0; i < computers.length; i++) {
            if (visited[i]) {
                continue;
            }
 
            stack.push(i);
            answer++;
 
            while(!stack.isEmpty()) {
                int cur = stack.pop();
 
                if (visited[cur]) {
                    continue;
                }
 
                visited[cur] = true;
 
                int[] computer = computers[cur];
 
                for (int j = 0; j < computer.length; j++) {
                    if (visited[j] || computer[j] == 0){
                        continue;
                    }
 
                    stack.push(j);
                }
            }
        }
 
 
        return answer;
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.py
def solution(n, computers):
    answer = 0
    check = [False] * n
    for i in range(0,n):
        if not check[i]:
            dfs(check, computers, i)
            answer += 1
        
    return answer

def dfs(check, computers, start):
    check[start] = True
    for i in range(0,len(computers)):
        if computers[start][i] == 1 and not check[i]:
            dfs(check, computers, i)
    